### PR TITLE
feat: Add Ethena integration

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

**Review by deHB6 — CHANGES_REQUESTED** [review]

File: src/MainnetController.sol [hunk 1]
```diff
@@ -112,10 +131,14 @@ contract MainnetController is AccessControl {
         daiUsds    = IDaiUsdsLike(daiUsds_);
         cctp       = ICCTPLike(cctp_);
 
-        susds = ISUSDSLike(susds_ );
+        ethenaMinter = IEthenaMinterLike(Ethereum.ETHENA_MINTER);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:134]: We built this system to do frequent updates of the controller so this isn't a problem. We would need advance notice and ideally both minter contracts run in parallel for a bit as we move over.

File: test/mainnet-fork/4626Calls.t.sol [hunk 2]
```diff
@@ -93,34 +93,34 @@ contract MainnetControllerDepositToSUSDSTests is SUSDSTestBase {
...
 
-contract MainnetControllerWithdrawFromSUSDSTests is SUSDSTestBase {
+contract MainnetControllerWithdrawERC4626Tests is SUSDSTestBase {
 
-    function test_withdrawFromSUSDS() external {
+    function test_withdrawERC4626() external {
         vm.startPrank(relayer);
         mainnetController.mintUSDS(1e18);
-        mainnetController.depositToSUSDS(1e18);
+        mainnetController.depositERC4626(address(susds), 1e18);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/mainnet-fork/4626Calls.t.sol:123]: I think we are okay with the way it currently is. If you guys can give us heads up with this sort of change we can take a look if/when that happens. As you say it is unlikely and if for some reason the standard logic for 4626 is broken we can always make code changes then.

File: src/MainnetController.sol [hunk 3]
```diff
@@ -203,56 +226,124 @@ contract MainnetController is AccessControl {
...
+            abi.encodeCall(susde.cooldownAssets, (usdeAmount))
+        );
+    }
+
+    function cooldownSharesSUSDe(uint256 susdeAmount) external onlyRole(RELAYER) isActive {
+        proxy.doCall(
+            address(susde),
+            abi.encodeCall(susde.cooldownShares, (susdeAmount))
+        );
+    }
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:338]: This is a fair point. This opportunity cost from griefing is large enough that we need to put a limit here.

@lucas-manuel let's add rate limits here. Both should be denominated in the USDe amount and share the same rate limit.

File: src/MainnetController.sol [hunk 4]
```diff
@@ -203,56 +226,124 @@ contract MainnetController is AccessControl {
...
+
+    function removeDelegatedSigner(address delegatedSigner) external onlyRole(RELAYER) isActive {
+        proxy.doCall(
+            address(ethenaMinter),
+            abi.encodeCall(ethenaMinter.removeDelegatedSigner, (address(delegatedSigner)))
+        );
+    }
+
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:306]: We discussed this, but didn't see any issues with having the relayer do it. Is there some security issue we should be concerned with?

The benefit of having the relayer do it is we can do something like rotate keys without the overhead of Sky governance process.

File: src/MainnetController.sol [hunk 5]
```diff
@@ -112,10 +131,14 @@ contract MainnetController is AccessControl {
...
 
-        susds = ISUSDSLike(susds_ );
+        ethenaMinter = IEthenaMinterLike(Ethereum.ETHENA_MINTER);
+
+        susde = ISUSDELike(Ethereum.SUSDE);
+        susds = ISUSDSLike(susds_);
         dai   = IERC20(daiUsds.dai());
         usdc  = IERC20(psm.gem());
         usds  = IERC20(susds.usds());
+        usde  = IERC20(Ethereum.USDE);
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:141]: What do you mean? The addresses are defined in our address registry.

File: src/MainnetController.sol [hunk 6]
```diff
@@ -203,56 +226,124 @@ contract MainnetController is AccessControl {
...
-        external onlyRole(RELAYER) isActive returns (uint256 shares)
+    function depositERC4626(address token, uint256 amount)
+        external
+        onlyRole(RELAYER)
+        isActive
+        rateLimited(
+            RateLimitHelpers.makeAssetKey(LIMIT_4626_DEPOSIT, token),  // TODO: Use makeAssetKey?
+            amount
+        )
+        returns (uint256 shares)
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/MainnetController.sol:240]: Ya this is fine - this is a backend piece.

File: src/MainnetController.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/MainnetController.sol:134]: Agree with Sam here, would just want ample notice if this contract were to change @deHB6, can update this contract to account for that

File: src/MainnetController.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/MainnetController.sol:306]: @deHB6 yeah can you expand on this thought further? If theres any security considerations related to having this as the RELAYER will be critical to change

File: src/MainnetController.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/MainnetController.sol:338]: Ok yeah lets rate limit this cooldown, and then in the situation where the cooldown is continuously refreshed so funds are "locked", we would freeze the controller and then replace the RELAYER with governance.

File: test/mainnet-fork/4626Calls.t.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/mainnet-fork/4626Calls.t.sol:123]: Yeah we also covered the sUSDe 4626 logic in the e2e tests so no worries

File: src/MainnetController.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (deHB6) [reply src/MainnetController.sol:306]: makes sense, esp if governance is a big pain/process

**Review by hexonaut — APPROVED** [review]

**Pull request merged**